usage_example/Readme.txt

```diff
@@ -0,0 +1,270 @@
+switched on based on previous config settings:
+
+CONFIG_HAVE_DMA_CONTIGUOUS=y
+CONFIG_ARCH_HAS_ZONE_DMA_SET=y
+CONFIG_DMA_SHARED_BUFFER=y
+CONFIG_DMA_ENGINE=y
+CONFIG_DMA_VIRTUAL_CHANNELS=y
+CONFIG_DMA_ACPI=y
+CONFIG_DMA_OF=y
+CONFIG_IOMMU_DEFAULT_DMA_LAZY=y
+CONFIG_IOMMU_DMA=y
+CONFIG_HAS_DMA=y
+CONFIG_DMA_OPS=y
+CONFIG_NEED_SG_DMA_FLAGS=y
+CONFIG_NEED_SG_DMA_LENGTH=y
+CONFIG_NEED_DMA_MAP_STATE=y
+CONFIG_ARCH_DMA_ADDR_T_64BIT=y
+CONFIG_DMA_DECLARE_COHERENT=y
+CONFIG_ARCH_HAS_SETUP_DMA_OPS=y
+CONFIG_ARCH_HAS_TEARDOWN_DMA_OPS=y
+CONFIG_ARCH_HAS_SYNC_DMA_FOR_DEVICE=y
+CONFIG_ARCH_HAS_SYNC_DMA_FOR_CPU=y
+CONFIG_ARCH_HAS_DMA_PREP_COHERENT=y
+CONFIG_DMA_BOUNCE_UNALIGNED_KMALLOC=y
+CONFIG_DMA_NONCOHERENT_MMAP=y
+CONFIG_DMA_COHERENT_POOL=y
+CONFIG_DMA_DIRECT_REMAP=y
+
+# CONFIG_DMABUF_HEAPS_CMA is not set
+
+
+The primary advantage of the CMA based approach over reserved memory is: 
+physical memory reserved by CMA can be utilized by Linux kernel for other 
+purposes (page/buffer cache) until it is claimed by DMA apis, after which 
+it is no longer used for other purposes. CMA memory is not used by kernel 
+memory allocators (SLAB/SLUB etc.) at any point. 
+
+Reserved physical memory is not used by kernel and has to be explicitly 
+configured to be utilized by modules/drivers. 
+
+Passing 'cma=xxx' as part of kernel command line reserves the necessary 
+amount of memory for CMA if kernel is configured with CMA feature enabled.
+
+[The kernel config option in the earlier step could be changed to reserve
+the necessary amount of CMA, but this may not be the right approach]
+
+Examples:
+AWS setup grub command line addition:
+
+menuentry 'Ubuntu, with Linux 6.5.13 cma' --class ubuntu --class gnu-linux --class gnu --class os $menuentry_id_option 'gnulinux-6.5.13-advanced-b993fcf5-2d55-493c-b232-97ca4539f4e9' {
+	recordfail
+	load_video
+	gfxmode $linux_gfx_mode
+	insmod gzio
+	if [ x$grub_platform = xxen ]; then insmod xzio; insmod lzopio; fi
+	insmod part_gpt
+	insmod ext2
+	search --no-floppy --fs-uuid --set=root b993fcf5-2d55-493c-b232-97ca4539f4e9
+	echo	'Loading Linux 6.5.13 ...'
+	if [ "${initrdfail}" = 1 ]; then
+	  echo	'GRUB_FORCE_PARTUUID set, initrdless boot failed. Attempting with initrd.'
+	  linux	/boot/vmlinuz-6.5.13 root=PARTUUID=26f22d2b-892e-459e-ba29-0fdab351f296 ro  console=tty1 console=ttyS0 cma=8G nvme_core.io_timeout=4294967295
+	  echo	'Loading initial ramdisk ...'
+	  initrd	/boot/initrd.img-6.5.13
+	else
+	  echo	'GRUB_FORCE_PARTUUID set, attempting initrdless boot.'
+	  linux	/boot/vmlinuz-6.5.13 root=PARTUUID=26f22d2b-892e-459e-ba29-0fdab351f296 ro  console=tty1 console=ttyS0 cma=8G nvme_core.io_timeout=4294967295 panic=-1
+	fi
+	initrdfail
+}
+
+
+In FPGA setup, setting Uboot command line by modifying the nandfitboot env variable: 
+
+setenv nandfitboot "setenv bootargs earlycon panic=-1 root=${nandroot} rw rootwait rootfstype=ubifs ubi.mtd=1 cma=256M;  bootm ${loadaddr} "
```

Comment:
Do we need to do this if the 512 MB is already reserved in DTS file in Linux ? How does one use that memory for this ?

Reply:
I have to test the module with using reserved memory, will do that, it should work. But if that reserved memory is also being used by other components directly accessing it through /dev/mem, I think it will lead to memory corruption issues. Creating another reserved memory for explicit contiguous memory usage should work and I can test that.

Reply:
Ok, I am approving the changes as is but lets test these changes to make sure our assumptions are correct.